eZ/Publish/Core/Repository/RoleService.php

```diff
@@ -157,23 +165,333 @@ public function createRole(APIRoleCreateStruct $roleCreateStruct)
+            );
+        } catch (APINotFoundException $e) {
+            $this->repository->beginTransaction();
+            try {
+                $spiRole = $this->userHandler->createRoleDraft($role->id);
+                $this->repository->commit();
+            } catch (Exception $e) {
+                $this->repository->rollback();
+                throw $e;
+            }
+        }
+
+        return $this->roleDomainMapper->buildDomainRoleDraftObject($spiRole);
+    }
+
+    /**
+     * Loads a RoleDraft for the given id.
+     *
+     * @since 6.0
+     *
+     * @throws \eZ\Publish\API\Repository\Exceptions\UnauthorizedException if the authenticated user is not allowed to read this RoleDraft
+     * @throws \eZ\Publish\API\Repository\Exceptions\NotFoundException if a RoleDraft with the given id was not found
+     *
+     * @param mixed $id
+     *
+     * @return \eZ\Publish\API\Repository\Values\User\RoleDraft
+     */
+    public function loadRoleDraft($id)
+    {
+        if ($this->repository->hasAccess('role', 'read') !== true) {
+            throw new UnauthorizedException('role', 'read');
+        }
+
+        $spiRole = $this->userHandler->loadRole($id, Role::STATUS_DRAFT);
+
+        return $this->roleDomainMapper->buildDomainRoleDraftObject($spiRole);
+    }
+
+    /**
+     * Updates the properties of a RoleDraft.
+     *
+     * @since 6.0
+     *
+     * @throws \eZ\Publish\API\Repository\Exceptions\UnauthorizedException if the authenticated user is not allowed to update a RoleDraft
+     * @throws \eZ\Publish\API\Repository\Exceptions\InvalidArgumentException if the identifier of the RoleDraft already exists
+     *
+     * @param \eZ\Publish\API\Repository\Values\User\RoleDraft $roleDraft
+     * @param \eZ\Publish\API\Repository\Values\User\RoleUpdateStruct $roleUpdateStruct
+     *
+     * @return \eZ\Publish\API\Repository\Values\User\RoleDraft
+     */
+    public function updateRoleDraft(APIRoleDraft $roleDraft, RoleUpdateStruct $roleUpdateStruct)
+    {
+        if ($roleUpdateStruct->identifier !== null && !is_string($roleUpdateStruct->identifier)) {
+            throw new InvalidArgumentValue('identifier', $roleUpdateStruct->identifier, 'RoleUpdateStruct');
+        }
+
+        $loadedRoleDraft = $this->loadRoleDraft($roleDraft->id);
+
+        if ($this->repository->hasAccess('role', 'update') !== true) {
+            throw new UnauthorizedException('role', 'update');
+        }
+
+        if ($roleUpdateStruct->identifier !== null) {
+            try {
+                /* Throw exception if:
+                 * - A published role with the same identifier exists, AND
+                 * - The ID of the two are NOT the same (otherwise it's not an editing conflict)
+                */
+                $existingRole = $this->loadRoleByIdentifier($roleUpdateStruct->identifier);
+                if ($existingRole instanceof APIRole && $existingRole->id != $loadedRoleDraft->id) {
```

Comment:
It is not necessary to check for the `instanceof`, we can rely on `loadRoleByIdentifier()` returning a Role or throwing an exception.